<a href="https://colab.research.google.com/github/Viny2030/UMSA-Analisis_de_datos/blob/main/EDa_completo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Cargo librerias

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

##Cargo Dataset

In [2]:
url = "https://raw.githubusercontent.com/Viny2030/UMSA-Analisis_de_datos/refs/heads/main/data.csv"

In [3]:
df = pd.read_csv(url)
df

,Make,Model,Year,Engine Fuel Type,Engine HP,Engine Cylinders,Transmission Type,Driven_Wheels,Number of Doors,Market Category,Vehicle Size,Vehicle Style,highway MPG,city mpg,Popularity,MSRP
0,BMW,1 Series M,2011,premium unleaded (required),335.0,6.0,MANUAL,rear wheel drive,2.0,"Factory Tuner,Luxury,High-Performance",Compact,Coupe,26,19,3916,46135
1,BMW,1 Series,2011,premium unleaded (required),300.0,6.0,MANUAL,rear wheel drive,2.0,"Luxury,Performance",Compact,Convertible,28,19,3916,40650
2,BMW,1 Series,2011,premium unleaded (required),300.0,6.0,MANUAL,rear wheel drive,2.0,"Luxury,High-Performance",Compact,Coupe,28,20,3916,36350
3,BMW,1 Series,2011,premium unleaded (required),230.0,6.0,MANUAL,rear wheel drive,2.0,"Luxury,Performance",Compact,Coupe,28,18,3916,29450
4,BMW,1 Series,2011,premium unleaded (required),230.0,6.0,MANUAL,rear wheel drive,2.0,Luxury,Compact,Convertible,28,18,3916,34500
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
11909,Acura,ZDX,2012,premium unleaded (required),300.0,6.0,AUTOMATIC,all wheel drive,4.0,"Crossover,Hatchback,Luxury",Midsize,4dr Hatchback,23,16,204,46120
11910,Acura,ZDX,2012,premium unleaded (required),300.0,6.0,AUTOMATIC,all wheel drive,4.0,"Crossover,Hatchback,Luxury",Midsize,4dr Hatchback,23,16,204,56670
11911,Acura,ZDX,2012,premium unleaded (required),300.0,6.0,AUTOMATIC,all wheel drive,4.0,"Crossover,Hatchback,Luxury",Midsize,4dr Hatchback,23,16,204,50620
11912,Acura,ZDX,2013,premium unleaded (recommended),300.0,6.0,AUTOMATIC,all wheel drive,4.0,"Crossover,Hatchback,Luxury",Midsize,4dr Hatchback,23,16,204,50920


In [4]:
df.shape

(11914, 16)

### Verificar tipo de dato

In [5]:
df.dtypes

,0
Make,object
Model,object
Year,int64
Engine Fuel Type,object
Engine HP,float64
Engine Cylinders,float64
Transmission Type,object
Driven_Wheels,object
Number of Doors,float64
Market Category,object


#### Renombrar columnas

In [6]:
df.rename(columns={'Make': 'Fabricante', 'Model': 'Modelo', 'Year': 'Año', 'Engine Fuel Type': 'Tipo_Combustible', 'Engine HP': 'Potencia_CV', 'Engine Cylinders': 'Cilindros_Motor', 'Transmission Type': 'Tipo_Transmision', 'Driven_Wheels': 'Ruedas_Motrices', 'Number of Doors': 'Numero_Puertas', 'Market Category': 'Categoria_Mercado', 'Vehicle Size': 'Tamaño_Vehiculo', 'Vehicle Style': 'Estilo_Vehiculo', 'highway MPG': 'MPG_Carretera', 'city mpg': 'MPG_Ciudad', 'Popularity': 'Popularidad', 'MSRP': 'Precio_Sugerido'}, inplace=True)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Fabricante         11914 non-null  object 
 1   Modelo             11914 non-null  object 
 2   Año                11914 non-null  int64  
 3   Tipo_Combustible   11911 non-null  object 
 4   Potencia_CV        11845 non-null  float64
 5   Cilindros_Motor    11884 non-null  float64
 6   Tipo_Transmision   11914 non-null  object 
 7   Ruedas_Motrices    11914 non-null  object 
 8   Numero_Puertas     11908 non-null  float64
 9   Categoria_Mercado  8172 non-null   object 
 10  Tamaño_Vehiculo    11914 non-null  object 
 11  Estilo_Vehiculo    11914 non-null  object 
 12  MPG_Carretera      11914 non-null  int64  
 13  MPG_Ciudad         11914 non-null  int64  
 14  Popularidad        11914 non-null  int64  
 15  Precio_Sugerido    11914 non-null  int64  
dtypes: float64(3), int64(5

### Planteo del Problema de Negocio

**Problema de Regresión: Predicción de Precio**
El objetivo principal de este conjunto de datos es predecir el precio sugerido de venta al público (Precio_Sugerido) de un vehículo basándose en sus características.

**Tipo de Aprendizaje: Supervisado.**

**Variable Objetivo (Target): Precio_Sugerido (MSRP)**. Esta es una variable continua (de tipo int64 y con un amplio rango de valores), lo que define el problema como de Regresión.

Características (Features): Todas las demás columnas, como Fabricante, Modelo, Potencia_CV, MPG_Carretera, Tipo_Transmision, etc.

**Formulación del Problema**
Se busca construir un modelo que pueda estimar con la mayor precisión posible el valor de Precio_Sugerido para cualquier coche nuevo o no visto en el conjunto de entrenamiento, basándose en un conjunto de características técnicas y de mercado.

#### Borrar columnas irrelevantes

In [8]:
df = df.drop(['Tipo_Combustible', 'Categoria_Mercado', 'Estilo_Vehiculo', 'Popularidad', 'Numero_Puertas', 'Tamaño_Vehiculo'], axis=1)
df.head(5)

,Fabricante,Modelo,Año,Potencia_CV,Cilindros_Motor,Tipo_Transmision,Ruedas_Motrices,MPG_Carretera,MPG_Ciudad,Precio_Sugerido
0,BMW,1 Series M,2011,335.0,6.0,MANUAL,rear wheel drive,26,19,46135
1,BMW,1 Series,2011,300.0,6.0,MANUAL,rear wheel drive,28,19,40650
2,BMW,1 Series,2011,300.0,6.0,MANUAL,rear wheel drive,28,20,36350
3,BMW,1 Series,2011,230.0,6.0,MANUAL,rear wheel drive,28,18,29450
4,BMW,1 Series,2011,230.0,6.0,MANUAL,rear wheel drive,28,18,34500


In [9]:
df

,Fabricante,Modelo,Año,Potencia_CV,Cilindros_Motor,Tipo_Transmision,Ruedas_Motrices,MPG_Carretera,MPG_Ciudad,Precio_Sugerido
0,BMW,1 Series M,2011,335.0,6.0,MANUAL,rear wheel drive,26,19,46135
1,BMW,1 Series,2011,300.0,6.0,MANUAL,rear wheel drive,28,19,40650
2,BMW,1 Series,2011,300.0,6.0,MANUAL,rear wheel drive,28,20,36350
3,BMW,1 Series,2011,230.0,6.0,MANUAL,rear wheel drive,28,18,29450
4,BMW,1 Series,2011,230.0,6.0,MANUAL,rear wheel drive,28,18,34500
...,...,...,...,...,...,...,...,...,...,...
11909,Acura,ZDX,2012,300.0,6.0,AUTOMATIC,all wheel drive,23,16,46120
11910,Acura,ZDX,2012,300.0,6.0,AUTOMATIC,all wheel drive,23,16,56670
11911,Acura,ZDX,2012,300.0,6.0,AUTOMATIC,all wheel drive,23,16,50620
11912,Acura,ZDX,2013,300.0,6.0,AUTOMATIC,all wheel drive,23,16,50920


### 1-Eliminar los duplicados-verificar

In [10]:
df.shape

(11914, 10)

In [11]:
duplicate_rows_df = df[df.duplicated()]
print("number of duplicate rows: ", duplicate_rows_df.shape)

number of duplicate rows:  (989, 10)


In [12]:
df.count()      # Used to count the number of rows

,0
Fabricante,11914
Modelo,11914
Año,11914
Potencia_CV,11845
Cilindros_Motor,11884
Tipo_Transmision,11914
Ruedas_Motrices,11914
MPG_Carretera,11914
MPG_Ciudad,11914
Precio_Sugerido,11914


In [13]:
df = df.drop_duplicates()
df.head(5)

,Fabricante,Modelo,Año,Potencia_CV,Cilindros_Motor,Tipo_Transmision,Ruedas_Motrices,MPG_Carretera,MPG_Ciudad,Precio_Sugerido
0,BMW,1 Series M,2011,335.0,6.0,MANUAL,rear wheel drive,26,19,46135
1,BMW,1 Series,2011,300.0,6.0,MANUAL,rear wheel drive,28,19,40650
2,BMW,1 Series,2011,300.0,6.0,MANUAL,rear wheel drive,28,20,36350
3,BMW,1 Series,2011,230.0,6.0,MANUAL,rear wheel drive,28,18,29450
4,BMW,1 Series,2011,230.0,6.0,MANUAL,rear wheel drive,28,18,34500


In [14]:
df.count()  ### cada vez que hago una tarea de preprocesamiento de datos debo controlar el df en ese punto de proceso

,0
Fabricante,10925
Modelo,10925
Año,10925
Potencia_CV,10856
Cilindros_Motor,10895
Tipo_Transmision,10925
Ruedas_Motrices,10925
MPG_Carretera,10925
MPG_Ciudad,10925
Precio_Sugerido,10925


In [15]:
df

,Fabricante,Modelo,Año,Potencia_CV,Cilindros_Motor,Tipo_Transmision,Ruedas_Motrices,MPG_Carretera,MPG_Ciudad,Precio_Sugerido
0,BMW,1 Series M,2011,335.0,6.0,MANUAL,rear wheel drive,26,19,46135
1,BMW,1 Series,2011,300.0,6.0,MANUAL,rear wheel drive,28,19,40650
2,BMW,1 Series,2011,300.0,6.0,MANUAL,rear wheel drive,28,20,36350
3,BMW,1 Series,2011,230.0,6.0,MANUAL,rear wheel drive,28,18,29450
4,BMW,1 Series,2011,230.0,6.0,MANUAL,rear wheel drive,28,18,34500
...,...,...,...,...,...,...,...,...,...,...
11909,Acura,ZDX,2012,300.0,6.0,AUTOMATIC,all wheel drive,23,16,46120
11910,Acura,ZDX,2012,300.0,6.0,AUTOMATIC,all wheel drive,23,16,56670
11911,Acura,ZDX,2012,300.0,6.0,AUTOMATIC,all wheel drive,23,16,50620
11912,Acura,ZDX,2013,300.0,6.0,AUTOMATIC,all wheel drive,23,16,50920


In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10925 entries, 0 to 11913
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Fabricante        10925 non-null  object 
 1   Modelo            10925 non-null  object 
 2   Año               10925 non-null  int64  
 3   Potencia_CV       10856 non-null  float64
 4   Cilindros_Motor   10895 non-null  float64
 5   Tipo_Transmision  10925 non-null  object 
 6   Ruedas_Motrices   10925 non-null  object 
 7   MPG_Carretera     10925 non-null  int64  
 8   MPG_Ciudad        10925 non-null  int64  
 9   Precio_Sugerido   10925 non-null  int64  
dtypes: float64(2), int64(4), object(4)
memory usage: 938.9+ KB


### Analisis columnas con faltantes

In [17]:
df.Potencia_CV.value_counts()

,count
Potencia_CV,
200.0,375
170.0,255
240.0,248
285.0,246
210.0,243
...,...
151.0,1
573.0,1
670.0,1


In [18]:
# Cuenta el número de valores que son NaN en la columna
df['Potencia_CV'].isnull().sum()

np.int64(69)

###Imputar faltantes con la media

In [19]:
# Si 'df' fue creado como una porción de otro DF, usa .loc:
df.loc[:, 'Potencia_CV'] = df['Potencia_CV'].fillna(df['Potencia_CV'].mean())

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10925 entries, 0 to 11913
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Fabricante        10925 non-null  object 
 1   Modelo            10925 non-null  object 
 2   Año               10925 non-null  int64  
 3   Potencia_CV       10925 non-null  float64
 4   Cilindros_Motor   10895 non-null  float64
 5   Tipo_Transmision  10925 non-null  object 
 6   Ruedas_Motrices   10925 non-null  object 
 7   MPG_Carretera     10925 non-null  int64  
 8   MPG_Ciudad        10925 non-null  int64  
 9   Precio_Sugerido   10925 non-null  int64  
dtypes: float64(2), int64(4), object(4)
memory usage: 938.9+ KB


In [21]:
df.Cilindros_Motor.value_counts()

,count
Cilindros_Motor,
4.0,4236
6.0,4231
8.0,1889
12.0,228
5.0,159
10.0,65
0.0,56
3.0,28
16.0,3


In [22]:
# Cuenta el número de valores que son NaN en la columna
df['Cilindros_Motor'].isnull().sum()

np.int64(30)

In [23]:
# Si 'df' fue creado como una porción de otro DF, usa .loc:
df.loc[:, 'Cilindros_Motor'] = df['Cilindros_Motor'].fillna(df['Cilindros_Motor'].mean())

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10925 entries, 0 to 11913
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Fabricante        10925 non-null  object 
 1   Modelo            10925 non-null  object 
 2   Año               10925 non-null  int64  
 3   Potencia_CV       10925 non-null  float64
 4   Cilindros_Motor   10925 non-null  float64
 5   Tipo_Transmision  10925 non-null  object 
 6   Ruedas_Motrices   10925 non-null  object 
 7   MPG_Carretera     10925 non-null  int64  
 8   MPG_Ciudad        10925 non-null  int64  
 9   Precio_Sugerido   10925 non-null  int64  
dtypes: float64(2), int64(4), object(4)
memory usage: 938.9+ KB


###

In [25]:
df

,Fabricante,Modelo,Año,Potencia_CV,Cilindros_Motor,Tipo_Transmision,Ruedas_Motrices,MPG_Carretera,MPG_Ciudad,Precio_Sugerido
0,BMW,1 Series M,2011,335.0,6.0,MANUAL,rear wheel drive,26,19,46135
1,BMW,1 Series,2011,300.0,6.0,MANUAL,rear wheel drive,28,19,40650
2,BMW,1 Series,2011,300.0,6.0,MANUAL,rear wheel drive,28,20,36350
3,BMW,1 Series,2011,230.0,6.0,MANUAL,rear wheel drive,28,18,29450
4,BMW,1 Series,2011,230.0,6.0,MANUAL,rear wheel drive,28,18,34500
...,...,...,...,...,...,...,...,...,...,...
11909,Acura,ZDX,2012,300.0,6.0,AUTOMATIC,all wheel drive,23,16,46120
11910,Acura,ZDX,2012,300.0,6.0,AUTOMATIC,all wheel drive,23,16,56670
11911,Acura,ZDX,2012,300.0,6.0,AUTOMATIC,all wheel drive,23,16,50620
11912,Acura,ZDX,2013,300.0,6.0,AUTOMATIC,all wheel drive,23,16,50920


In [26]:
df.to_csv('data_clean.csv', index=False)

###Preparacion columnas string para la aplicacion del Algoritmo

##Analisis de las columnas

In [27]:
df.Fabricante.value_counts().count()

np.int64(48)

In [28]:
df.Modelo.value_counts().count()

np.int64(915)

In [29]:
df.Tipo_Transmision.value_counts().count()

np.int64(5)

In [30]:
df.Ruedas_Motrices.value_counts().count()

np.int64(4)

## Al contar los valores equivalen a las columnas de transformacion del metodo pd.get_dummies()

### Decision eliminar la columna MOdelo generaria  915 columnas realmente no es practico para aplicar el algoritmo.

## eliminacion de la columna Modelo

In [31]:
df = df.drop(['Modelo'], axis=1)

In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10925 entries, 0 to 11913
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Fabricante        10925 non-null  object 
 1   Año               10925 non-null  int64  
 2   Potencia_CV       10925 non-null  float64
 3   Cilindros_Motor   10925 non-null  float64
 4   Tipo_Transmision  10925 non-null  object 
 5   Ruedas_Motrices   10925 non-null  object 
 6   MPG_Carretera     10925 non-null  int64  
 7   MPG_Ciudad        10925 non-null  int64  
 8   Precio_Sugerido   10925 non-null  int64  
dtypes: float64(2), int64(4), object(3)
memory usage: 853.5+ KB


In [33]:
df = pd.get_dummies(df, columns=['Fabricante', 'Tipo_Transmision', 'Ruedas_Motrices'])

In [34]:
df

,Año,Potencia_CV,Cilindros_Motor,MPG_Carretera,MPG_Ciudad,Precio_Sugerido,Fabricante_Acura,Fabricante_Alfa Romeo,Fabricante_Aston Martin,Fabricante_Audi,...,Fabricante_Volvo,Tipo_Transmision_AUTOMATED_MANUAL,Tipo_Transmision_AUTOMATIC,Tipo_Transmision_DIRECT_DRIVE,Tipo_Transmision_MANUAL,Tipo_Transmision_UNKNOWN,Ruedas_Motrices_all wheel drive,Ruedas_Motrices_four wheel drive,Ruedas_Motrices_front wheel drive,Ruedas_Motrices_rear wheel drive
0,2011,335.0,6.0,26,19,46135,False,False,False,False,...,False,False,False,False,True,False,False,False,False,True
1,2011,300.0,6.0,28,19,40650,False,False,False,False,...,False,False,False,False,True,False,False,False,False,True
2,2011,300.0,6.0,28,20,36350,False,False,False,False,...,False,False,False,False,True,False,False,False,False,True
3,2011,230.0,6.0,28,18,29450,False,False,False,False,...,False,False,False,False,True,False,False,False,False,True
4,2011,230.0,6.0,28,18,34500,False,False,False,False,...,False,False,False,False,True,False,False,False,False,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
11909,2012,300.0,6.0,23,16,46120,True,False,False,False,...,False,False,True,False,False,False,True,False,False,False
11910,2012,300.0,6.0,23,16,56670,True,False,False,False,...,False,False,True,False,False,False,True,False,False,False
11911,2012,300.0,6.0,23,16,50620,True,False,False,False,...,False,False,True,False,False,False,True,False,False,False
11912,2013,300.0,6.0,23,16,50920,True,False,False,False,...,False,False,True,False,False,False,True,False,False,False


In [35]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10925 entries, 0 to 11913
Data columns (total 63 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Año                                10925 non-null  int64  
 1   Potencia_CV                        10925 non-null  float64
 2   Cilindros_Motor                    10925 non-null  float64
 3   MPG_Carretera                      10925 non-null  int64  
 4   MPG_Ciudad                         10925 non-null  int64  
 5   Precio_Sugerido                    10925 non-null  int64  
 6   Fabricante_Acura                   10925 non-null  bool   
 7   Fabricante_Alfa Romeo              10925 non-null  bool   
 8   Fabricante_Aston Martin            10925 non-null  bool   
 9   Fabricante_Audi                    10925 non-null  bool   
 10  Fabricante_BMW                     10925 non-null  bool   
 11  Fabricante_Bentley                 10925 non-null  bool   


### Ordenar la columna objetivo al final

In [36]:
# Mueve 'Precio_Sugerido' al final de la lista de columnas
df = df[[col for col in df.columns if col != 'Precio_Sugerido'] + ['Precio_Sugerido']]

In [37]:
df

,Año,Potencia_CV,Cilindros_Motor,MPG_Carretera,MPG_Ciudad,Fabricante_Acura,Fabricante_Alfa Romeo,Fabricante_Aston Martin,Fabricante_Audi,Fabricante_BMW,...,Tipo_Transmision_AUTOMATED_MANUAL,Tipo_Transmision_AUTOMATIC,Tipo_Transmision_DIRECT_DRIVE,Tipo_Transmision_MANUAL,Tipo_Transmision_UNKNOWN,Ruedas_Motrices_all wheel drive,Ruedas_Motrices_four wheel drive,Ruedas_Motrices_front wheel drive,Ruedas_Motrices_rear wheel drive,Precio_Sugerido
0,2011,335.0,6.0,26,19,False,False,False,False,True,...,False,False,False,True,False,False,False,False,True,46135
1,2011,300.0,6.0,28,19,False,False,False,False,True,...,False,False,False,True,False,False,False,False,True,40650
2,2011,300.0,6.0,28,20,False,False,False,False,True,...,False,False,False,True,False,False,False,False,True,36350
3,2011,230.0,6.0,28,18,False,False,False,False,True,...,False,False,False,True,False,False,False,False,True,29450
4,2011,230.0,6.0,28,18,False,False,False,False,True,...,False,False,False,True,False,False,False,False,True,34500
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
11909,2012,300.0,6.0,23,16,True,False,False,False,False,...,False,True,False,False,False,True,False,False,False,46120
11910,2012,300.0,6.0,23,16,True,False,False,False,False,...,False,True,False,False,False,True,False,False,False,56670
11911,2012,300.0,6.0,23,16,True,False,False,False,False,...,False,True,False,False,False,True,False,False,False,50620
11912,2013,300.0,6.0,23,16,True,False,False,False,False,...,False,True,False,False,False,True,False,False,False,50920


In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10925 entries, 0 to 11913
Data columns (total 63 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Año                                10925 non-null  int64  
 1   Potencia_CV                        10925 non-null  float64
 2   Cilindros_Motor                    10925 non-null  float64
 3   MPG_Carretera                      10925 non-null  int64  
 4   MPG_Ciudad                         10925 non-null  int64  
 5   Fabricante_Acura                   10925 non-null  bool   
 6   Fabricante_Alfa Romeo              10925 non-null  bool   
 7   Fabricante_Aston Martin            10925 non-null  bool   
 8   Fabricante_Audi                    10925 non-null  bool   
 9   Fabricante_BMW                     10925 non-null  bool   
 10  Fabricante_Bentley                 10925 non-null  bool   
 11  Fabricante_Bugatti                 10925 non-null  bool   


### Queda listo el dataset para la aplicacion del Algoritmo de Regresion